src/python/pants/util/dirutil.py

```diff
@@ -108,46 +123,77 @@ def safe_file_dump(filename, payload='', mode='w', makedirs=False):
   When `payload` is an empty string (the default), this method can be used as a concise way to
   create an empty file along with its containing directory (or truncate it if it already exists).
 
-  :param string filename: The filename of the file to write to.
-  :param string payload: The string to write to the file.
-  :param string mode: A mode argument for the python `open` builtin which should be a write mode
-                      variant. Defaults to 'w'.
-  :param bool makedirs: Whether to make all parent directories of this file before making it.
+  :param filename: The filename of the file to write to.
+  :param payload: The string to write to the file.
+  :param mode: A mode argument for the python `open` builtin which should be a write mode variant.
+               Defaults to 'w'.
+  :param makedirs: Whether to make all parent directories of this file before making it.
   """
   if makedirs:
     os.makedirs(os.path.dirname(filename), exist_ok=True)
   with safe_open(filename, mode=mode) as f:
     f.write(payload)
 
 
-def maybe_read_file(filename, binary_mode=False):
+@overload
+def maybe_read_file(filename: str) -> Optional[str]: ...
+
+
+@overload
+def maybe_read_file(filename: str, binary_mode: Literal[False]) -> Optional[str]: ...
```

Comment:
This uses two really cool features from mypy: `@overload` and `Literal`. It allows our type hints to get much more precise so that we don't always have to deal with `Union`.

Reply:
This is preeeetty magical. Not sure I love it, but I don't hate it.

Reply:
Agreed, although a really nice experience for the caller of this function to have the extra precision. 

Fortunately, we shouldn't have this type of overload frequently - its main use is when return type differs based on param types, and that's not very common to happen.